test/test_fork.py

```diff
@@ -106,10 +107,27 @@ def action(client):
                     rc = self.clients[i % len(self.clients)]
                     if i % 50 == 0 and self.fork:
                         # Fork
-                        with self.runner.fork() as pid:
-                            if pid == 0:  # Child
-                                for c in self.clients:
-                                    action(c)
+                        def target():
+                            for c in self.clients:
+                                action(c)
+
+                        import multiprocessing
+
+                        ctx = multiprocessing.get_context("fork")
+                        proc = ctx.Process(target=target)
+                        proc.start()
+                        # Wait 60s.
+                        proc.join(60)
+                        if proc.exitcode is None:
+                            # If it failed, SIGINT to get traceback and wait
+                            # 10s.
+                            os.kill(proc.pid, signal.SIGINT)
+                            proc.join(10)
+                            if proc.exitcode is None:
+                                # If that also failed, SIGKILL and resume after.
+                                os.kill(proc.pid, signal.SIGKILL)
+                            self.runner.fail("deadlock")
```

Comment:
Need one last `proc.join(10)` before failing here to avoid zombie process or a resource warning.